_posts/2024-11-22-faiss-byte-vector.md

```diff
@@ -0,0 +1,192 @@
+|**Dataset 3**	|cohere-ip-1m	|768	|1,000,000	|10,000	|[-4.1073565, 5.504557]	|[-4.109505, 5.4809895]	|innerproduct	|
+|**Dataset 4**	|sift-128-euclidean	|128	|1,000,000	|10,000	|[0.0, 218.0]	|[0.0, 184.0]	|L2	|
+
+### Recall, memory, and indexing results
+
+|Dataset ID	|Faiss HNSW recall@100	|Faiss HNSW-byte recall@100	|% Reduction in recall	|Faiss HNSW memory usage (GB)	|Faiss HNSW byte memory usage (GB)	|% Reduction in memory	|Faiss HNSW mean indexing throughput (docs/sec)	|Faiss HNSW byte mean indexing throughput (docs/sec)	|% Gain in indexing throughput	|
+|---	|---	|---	|---	|---	|---	|---	|---	|---	|---	|
+|**Dataset 1**	|0.91	|0.89	|2.20	|3.72	|1.04	|72.00	|4673	|9686	|107.28	|
+|**Dataset 2**	|0.91	|0.83	|8.79	|30.03	|8.57	|71.46	|4911	|10207	|107.84	|
+|**Dataset 3**	|0.94	|0.86	|8.51	|3.00	|0.86	|71.33	|6112	|11673	|90.98	|
+|**Dataset 4**	|0.99	|0.98	|1.01	|0.62	|0.26	|58.06	|38273	|43267	|13.05	|
+
+### Query results
+
+|Dataset ID	|Query clients	|Faiss HNSW p90 (ms)	|Faiss HNSW byte p90 (ms)	|Faiss HNSW p99 (ms)	|Faiss HNSW byte p99 (ms)	|
+|---	|---	|---	|---	|---	|---	|
+|**Dataset 1**	|**1**	|5.35	|5.34	|5.95	|5.59	|
+|**Dataset 1**	|**8**	|6.68	|6.64	|10.23	|9.14	|
+|**Dataset 1**	|**16**	|10.59	|7.38	|12.94	|11.47	|
+|	|	|	|	|	|	|
+|**Dataset 2**	|**1**	|7.39	|7.14	|8.35	|7.59	|
+|**Dataset 2**|**8**	|15.47	|14.83	|21.38	|16.20	|
+|**Dataset 2**	|**16**	|25.01	|25.32	|31.98	|29.42	|
+|	|	|	|	|	|	|
+|**Dataset 3**	|**1**	|4.97	|4.72	|5.62	|5.02	|
+|**Dataset 3**	|**8**	|6.75	|5.98	|7.69	|7.7	|
+|**Dataset 3**	|**16**	|10.51	|6.94	|13.87	|12.4	|
+|	|	|	|	|	|	|
+|**Dataset 4**	|**1**	|2.91	|3.03	|3.16	|3.15	|
+|**Dataset 4**|**8**	|3.38	|3.30	|6.30	|4.75	|
+|**Dataset 4**	|**16**	|4.35	|3.80	|8.76	|8.83	|
+
+### Key findings
+
+When comparing the benchmarking results, here are the key findings:
+
+- **Memory savings**: Byte vectors reduced memory usage by up to **72%**, with higher-dimensional vectors achieving greater reductions.
+- **Indexing performance**: The mean indexing throughput for byte vectors was **2x to 107.84%** higher than for float vectors, especially with larger vector dimensions.
+- **Search performance**: Search latencies were similar, with byte vectors occasionally performing better.
+- **Recall**: For byte vectors, there was a slight (up to **8.8%**) reduction in recall compared to float vectors, depending on the dataset and the quantization technique used.
+
+## How does Faiss work with byte vectors internally?
+
+Faiss doesn't directly support the `byte` data type for storing vectors. To achieve this, OpenSearch uses a  [`QT_8bit_direct_signed` scalar quantizer](https://faiss.ai/cpp_api/struct/structfaiss_1_1ScalarQuantizer.html). This quantizer accepts float vectors within the signed 8-bit value range and encodes them as unsigned 8-bit integer vectors. During indexing and search, these encoded unsigned 8-bit integer vectors are decoded back into signed 8-bit original vectors for distance computation.
+
```

Comment:
Line 160, end of last sentence: "are decoded back into the original signed 8-bit vectors for the purpose of distance computation"?

Reply:
I think it's understandable in the original form.

Reply:
It reads as though "the" should precede "signed", though.